srml/elections-phragmen/src/lib.rs

```diff
@@ -0,0 +1,957 @@
+	decl_module, dispatch,
+	traits::{Currency, Get, LockableCurrency, LockIdentifier, ReservableCurrency, WithdrawReason,
+		WithdrawReasons, ChangeMembers
+	}
+};
+use system::{self, ensure_signed, ensure_root};
+
+const MODULE_ID: LockIdentifier = *b"phrelect";
+
+type BalanceOf<T> = <<T as Trait>::Currency as Currency<<T as system::Trait>::AccountId>>::Balance;
+
+pub const DEFAULT_CANDIDACY_BOND: u32 = 9;
+pub const DEFAULT_VOTING_BOND: u32 = 2;
+
+pub trait Trait: system::Trait {
+	type Event: From<Event<Self>> + Into<<Self as system::Trait>::Event>;
+
+	/// The currency that people are electing with.
+	type Currency:
+		LockableCurrency<Self::AccountId, Moment=Self::BlockNumber> +
+		ReservableCurrency<Self::AccountId>;
+
+	/// What to do when the members change.
+	type ChangeMembers: ChangeMembers<Self::AccountId>;
+
+	/// Convert a balance into a number used for election calculation.
+	/// This must fit into a `u64` but is allowed to be sensibly lossy.
+	type CurrencyToVote: Convert<BalanceOf<Self>, u64> + Convert<u128, BalanceOf<Self>>;
+
+	/// How much should be locked up in order to submit one's candidacy.
+	type CandidacyBond: Get<BalanceOf<Self>>;
+
+	/// How much should be locked up in order to be able to submit votes.
+	type VotingBond: Get<BalanceOf<Self>>;
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as PhragmenElection {
+		// ---- parameters
+		/// Number of seats to elect.
+		pub DesiredSeats get(desired_seats) config(): u32;
+		/// The total number of vote rounds that have happened, exclusive of the upcoming one.
+		pub ElectionRounds get(election_rounds): u32 = Zero::zero();
+		/// How long each seat is kept. This defined the next block number at which an election
+		/// round will happen.
+		pub TermDuration get(term_duration) config(): T::BlockNumber;
+
+		// ---- State
+		/// The current elected membership. Sorted.
+		pub Members get(members) config(): Vec<T::AccountId>;
+
+		/// Votes of a particular voter, with the round index of the votes.
+		pub VotesOf get(votes_of): linked_map T::AccountId => (Vec<T::AccountId>, u32);
+		/// Locked stake of a voter.
+		pub StakeOf get(stake_of): map T::AccountId => BalanceOf<T>;
+
+		/// The present candidate list. Always sorted.
+		pub Candidates get(candidates): Vec<T::AccountId>;
+		/// Current number of active candidates.
+		pub CandidateCount get(candidate_count): u32;
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		fn deposit_event<T>() = default;
+
+		const CandidacyBond: BalanceOf<T> = T::CandidacyBond::get();
+		const VotingBond: BalanceOf<T> = T::VotingBond::get();
+
+		/// Vote for a set of candidates for the upcoming round of election.
+		///
+		/// The `votes` should:
+		///   - not be empty.
+		///   - not be less than the number of candidates.
+		///
+		/// `who` cannot be an old voter who has not yet called [`remove_voter`].
```

Comment:
```suggestion
		/// `who` cannot be an existing voter
```

Reply:
actually: it _can be an existing voter_ to _update_ their vote. 

It cannot be, exactly as phrased: _old voter who has not yet called `remove_voter`_

Reply:
is "old voter" a technical term?

Reply:
The full story is: 

The code does not clean the votes at all and relies on two incentivization that voters themselves will do it. As a voter, once your vote has been counted once, then you are an _old voter_  aka. retired if you want to name it like that. You should manually remove yourself via another tx because:

1. you want your reserved bond back.
2. you can't vote and participate anymore in rounds.

If you are fine with the logic I can formulate it better as _inactive voter_  is the module docs. I agree that the term _old voter_ is not clear.

Reply:
`defunct_voter`